### PR TITLE
greendns: Return answers from /etc/hosts despite nameserver errors

### PR DESCRIPTION
**Pull request by tipabu**
When calling `getaddrinfo` with an unspecified address family, we:

* try to get an IPv6 address for the host, first trying /etc/hosts and failing that, using any external nameserver, then
* try to get an IPv4 address for it, with the same /etc/hosts -> external nameserver priority.

Note that if the host can be resolved using /etc/hosts but only to one of the two address families, we still attempt to contact the external nameserver to resolve it in the other family.

Previously, if this timed out or raised a `DNSException` other than `NXDOMAIN`, `EAI_EAGAIN_ERROR` or `EAI_NODATA_ERROR` would be raised even though we could resolve the host. Now, those errors will only be raised if neither attempt at resolution succeeded.

### CODE REVIEW

File: eventlet/support/greendns.py [hunk 1]
```diff
@@ -402,10 +402,24 @@ def _getaddrinfo_lookup(host, family, flags):
...
+            except socket.gaierror as e:
+                if e.errno == socket.EAI_AGAIN:
+                    EAGAIN_caught = True
+                elif e.errno == socket.EAI_NODATA:
+                    NODATA_caught = True
+                else:
+                    raise
+            else:
+                if answer.rrset:
+                    addrs.extend([rr.address for rr in answer.rrset])
```
Comment (temoto) [inline eventlet/support/greendns.py]: `list.extend` takes iterable. If there were no good reason for materialising list argument, then `extend(rr.a for rr in rrset)` is equally good.

File: eventlet/support/greendns.py [hunk 2]
```diff
@@ -402,10 +402,24 @@ def _getaddrinfo_lookup(host, family, flags):
...
+        if EAGAIN_caught and not addrs:
+            raise EAI_EAGAIN_ERROR
```
Comment (temoto) [inline eventlet/support/greendns.py]: Saving original exception seems better than raising a new one. Like

```
err = None
try:
  ...
except socket.gaierror as e:
  if e.errno in (again, nodata):
    err = e
...
if err is not None and not addrs:
  raise err
```

**Comment by tipabu** [comment]
Good call on the `extend` call; I was just moving code, but that makes it much better.

I also fixed up the error trapping. Seems much cleaner, good recommendation.

**Comment by temoto** [comment]
Thank you.

**Pull request merged**